### ISSUE TITLE
[autoscaler] negative CPU reservation breaks auto-scaling if a remote function spawns additional remote functions and fetches their results

### ISSUE THREAD

**Issue by sjakthol** [ray-project/ray#4743]
### System information
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Linux Ubuntu 18.04
- **Ray installed from (source or binary)**: Binary
- **Ray version**: 0.6.6
- **Python version**: 3.6.7
- **Exact commands to reproduce**:
  1. Install Ray: `pip install ray==0.6.6`
  2. Download sample cluster config: `curl -LO https://gist.github.com/sjakthol/b383f574380280cf5d8a6a62a9a6452e/raw/169cbd6e0aedfdc92e65d50e1cdd905574139563/ray-autoscaler-negative-cpu-autoscaling-config.yaml`(or visit https://gist.github.com/sjakthol/b383f574380280cf5d8a6a62a9a6452e)
  3. Download sample application: `curl -LO https://gist.github.com/sjakthol/b383f574380280cf5d8a6a62a9a6452e/raw/169cbd6e0aedfdc92e65d50e1cdd905574139563/ray-autoscaler-negative-cpu-sample-app.py` (or visit https://gist.github.com/sjakthol/b383f574380280cf5d8a6a62a9a6452e)
  4. Start Ray head with the dummy auto-scaling config: `ray stop && rm -r /tmp/ray/ && ray start --head --redis-port 6379 --autoscaling-config ray-autoscaler-negative-cpu-autoscaling-config.yaml`
  5. Execute the sample application: `python3 ray-autoscaler-negative-cpu-sample-app.py`
  6. Wait for the app to finish (takes around 10 secs)
  7. Check monitor.err and raylet.err for odd output (raylet oversubscription errors and negative resource reservation reports from the autoscaler).

Please note that the steps to reproduce abuse the local autoscaler backend a bit to make the issue a bit easier to reproduce. The problem was initially observed on a Ray cluster deployed to AWS with the AWS autoscaler backend and can be reproduced in that environment as well.

### Describe the problem
I have a Ray application in which remote functions spawn additional remote functions and do `ray.get()` to wait for their results. The application was executed on a Ray cluster deployed to AWS. When the execution finished, Ray autoscaler started to report a negative value for the number reserved CPUs in the cluster's `monitor.err` log file.

Upon further investigation, this does not appear to be just a cosmetic issue in the Ray log output. The negative CPU reservation also affected auto-scaling actions as the CPU reservation (and hence, cluster utilization) was negative (or in case of cluster utilization it was clamped to 0). Basically, the cluster scaled down to `min_nodes` and never scaled up again. The only recourse was to terminate the cluster and provision a new one.

### Source Code / Logs
Here are the key parts of the sample application used in the steps-to-reproduce above:

```python
import multiprocessing

import ray
ray.init(redis_address='localhost:6379')

NUM_CPUS = multiprocessing.cpu_count()

@ray.remote
def inner(i):
    return i + 1

@ray.remote
def outer(i):
    for i in range(100):
        ray.get(inner.remote(i))

ray.get([outer.remote(i) for i in range(2 * NUM_CPUS)])
```

Key points to note are:
* The app calls `outer.remote()` multiple times to ensure the Ray cluster is fully utilized.
* The `outer()` function calls a second Ray remote function, `inner()`, in a loop waiting for each invocation to finish before starting the next one

When this application is executed, the following messages are printed to `raylet.err`:
```
W0504 10:20:18.142570 19698 node_manager.cc:1543] Resources oversubscribed: {GPU,0.000000},{CPU,0.000000}
W0504 10:20:18.146426 19698 node_manager.cc:1543] Resources oversubscribed: {GPU,0.000000},{CPU,1.000000}
[...]
```

Once all the tasks of the app finish, `monitor.err` starts to report negative values for reserved CPU cores:
```
2019-05-04 10:20:22,337 INFO autoscaler.py:631 -- LoadMetrics: MostDelayedHeartbeats={'192.168.0.103': 0.10772442817687988}, NodeIdleSeconds=Min=0 Mean=0 Max=0, NumNodesConnected=1, NumNodesUsed=0.0, ResourceUsage=-659.0/4.0 b'CPU', 0.0/0.0 b'GPU', TimeSinceLastHeartbeat=Min=0 Mean=0 Max=0
```

If we submit additional applications to the cluster, the ResourceUsage value grows a bit but returns back to the original negative value (or go lower if the same issue was hit again). The NumNodesUsed stays at 0.0 forever. As far as I know, NumNodesUsed is the value the StandardAutoscaler class uses to determine if cluster needs to be scaled up or down. And if this value is stuck at 0.0, the cluster will eventually scale down and never scale up again.

The full source code for the sample app and cluster config I used to reproduce this issue can be found from [this Gist](https://gist.github.com/sjakthol/b383f574380280cf5d8a6a62a9a6452e). The Gist also contains the full `monitor.err` and `raylet.err` log files for execution of the steps listed in the beginning of this ticket.

### Investigation on Possible Cause(s)
The `raylet.err` resource oversubscription messages are logged at https://github.com/ray-project/ray/blob/c26f24ab9f45c8aea318d978107e116e7ecc9371/src/ray/raylet/node_manager.cc#L1563-L1567. From the surrounding code it looks like the issue is related to how remote functions release resources while they are blocked on a call to `ray.get()`. The code seems to be handling the case when the object(s) `ray.get()` is waiting on become available and the task resumes execution. This involves reclaiming the resources it had released earlier. Now, if the worker is already fully booked, the resuming task is unable to reacquire the required resource and continues execution without those. And this confuses the autoscaler resource reservation tracking.

Now, I haven't been able to trace how exactly does the autoscaler get the resource usage information. But if I had to guess, the chain of event that lead to this issue are:
1. The initial CPU reservation is 0.0 (other concurrent tasks are excluded from this number).
2. A remote function is started. It reserves one CPU core. CPU reservation goes to 1.0.
3. The remote function calls `ray.get()`. It releases resources. CPU reservation goes to 0.0.
4. The remote function is unblocked. It tries to reacquire one CPU core but cannot do that. CPU reservation is still 0.0.
5. The remote function call finishes. It releases the one CPU core everyone thought it would have. CPU reservation goes to -1.0.

In short, the component that monitors CPU reservation somehow double counts the released resources (once when blocked on `ray.get()` and second time when the remote function returns) as it assumed the remote function was able to reacquire them when it was unblocked.

Please let me know if you need more information to reproduce this issue.

**Comment by sjakthol** [comment]
I noticed that there has been bunch of changes to resource accounting done in NodeManager lately and tested a couple of other Ray versions as well to see if those changes had anything to do with in.
* The issue occurs with ray==0.6.5 as well. So https://github.com/ray-project/ray/commit/0f42f87ebc83c2541b0ba2c70042d2283e733dce that went into v0.6.6 did not introduce this issue.
* The Ray build from https://s3-us-west-2.amazonaws.com/ray-wheels/latest/ray-0.7.0.dev3-cp36-cp36m-manylinux1_x86_64.whl (downloaded at the time of writing) does not log anything about resource usage (the log just says `ResourceUsage=,` after running the PoC). So I cannot tell if the issue persists or not with the latest builds from master.

**Comment by robertnishihara** [comment]
Hi @sjakthol, thanks a lot for reporting this. Yes, currently there are brief periods of time, where the scheduler can have "negative" resource availability (when a tasks that released its CPU resources in a `ray.get` call reacquires those resources when the `ray.get` call returns).

The scheduler will broadcast load information in heartbeats, which eventually get processed by the autoscaler at https://github.com/ray-project/ray/blob/1a39fee9c6fa0f81e1184182b47ef106269613fe/python/ray/autoscaler/autoscaler.py#L159-L175

There may be an issue somewhere in https://github.com/ray-project/ray/blob/1a39fee9c6fa0f81e1184182b47ef106269613fe/python/ray/autoscaler/autoscaler.py#L214-L264

where it doesn't properly handle negative values.

A better fix would probably be to make sure we never broadcast negative values, e.g., by modifying

https://github.com/ray-project/ray/blob/1a39fee9c6fa0f81e1184182b47ef106269613fe/src/ray/raylet/node_manager.cc#L297-L305

to take a max with 0 before setting the available resources.

cc @hartikainen @richardliaw @ericl
